File: telephony/src/lib/telnyxClient.ts
// telephony/src/lib/telnyxClient.ts
// Converted from telnyx_client.go

import { getConfig } from "./config";

export function say(callControlID: string, text: string): void {
  const cfg = getConfig();
  sendCommand(callControlID, "actions/speak", {
    language: "en-US",
    voice: cfg.telnyxVoice,
    payload: text,
  });
}

export function sendCommand(
  callControlID: string,
  action: string,
  payload: Record<string, unknown> = {}
): void {
  const cfg = getConfig();
  const url = `https://api.telnyx.com/v2/calls/${callControlID}/${action}`;

  fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      Authorization: `Bearer ${cfg.telnyxAPIKey}`,
    },
    body: JSON.stringify(payload),
    signal: AbortSignal.timeout(10_000),
  })
    .then(async (resp) => {
      if (!resp.ok) {
        const body = await resp.text().catch(() => "");
        console.error("[telnyx] API error", { action, status: resp.status, body });
      } else {
        console.log("[telnyx] command sent", { action, callControlID });
      }
    })
    .catch((err) => {
      console.error("[telnyx] request failed", { action, err });
    });
}

File: telephony/_go_legacy/telnyx_client.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"time"
)

func say(callControlID, text string) {
	sendCommand(callControlID, "actions/speak", map[string]interface{}{
		"language": "en-US",
		"voice":    cfg.TelnyxVoice,
		"payload":  text,
	})
}

func sendCommand(callControlID, action string, payload map[string]interface{}) {
	if payload == nil {
		payload = map[string]interface{}{}
	}

	url := fmt.Sprintf("https://api.telnyx.com/v2/calls/%s/%s", callControlID, action)
	jsonData, err := json.Marshal(payload)
	if err != nil {
		slog.Error("Failed to marshal command payload", "action", action, "err", err)
		return
	}

	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(jsonData))
	if err != nil {
		slog.Error("Failed to create Telnyx request", "action", action, "err", err)
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Authorization", "Bearer "+cfg.TelnyxAPIKey)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		slog.Error("Telnyx API request failed", "action", action, "err", err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		slog.Error("Telnyx API error",
			"action", action,
			"status", resp.StatusCode,
			"body", string(respBody),
		)
		return
	}

	slog.Info("Telnyx command sent", "action", action, "call_id", callControlID)
}
